lib/services/app.go

```diff
@@ -158,3 +169,217 @@ func UnmarshalAppServer(data []byte, opts ...MarshalOption) (types.AppServer, er
 	}
 	return nil, trace.BadParameter("unsupported app server resource version %q", h.Version)
 }
+
+// NewApplicationsFromKubeService creates application resources from kubernetes service. One service result
+// in multiple application resources if there are multiple ports exposed.
+func NewApplicationsFromKubeService(service v1.Service, clusterName string, pi ProtocolChecker) ([]types.Application, error) {
```

Comment:
this function is not exactly trivial, and I notice most of the inner functions have unit tests but this outer function doesn't, is it possible to test the full `NewApplicationsFromKubeService`

Reply:
It's tested extensively a level higher in the next PR (it was originally one, before I split it into 3): https://github.com/gravitational/teleport/pull/28988/files#diff-8e8aba724ead483a6b4d48d816683b0793980d4badaeaa9e460b1fac5f73aa35